addons/website_sale/controllers/main.py

```diff
@@ -351,8 +353,12 @@ def cart(self, **post):
 
     @http.route(['/shop/cart/update'], type='http', auth="public", methods=['POST'], website=True)
     def cart_update(self, product_id, add_qty=1, set_qty=0, **kw):
-        request.website.sale_get_order(force_create=1)._cart_update(product_id=int(product_id), add_qty=float(add_qty), set_qty=float(set_qty))
-        return request.redirect("/shop/cart")
+        attr = {}
+        values = request.website.sale_get_order(force_create=1)._cart_update(product_id=int(product_id), add_qty=float(add_qty), set_qty=float(set_qty))
+        if 'message' in values:
+            attr = dict(message=values.get('message'), message_type=values.get('message_type'))
+            return request.redirect("/shop/cart%s%s" % (attr and "?" or "", attr and (werkzeug.url_encode(attr) or "")))
```

Comment:
* values.get('message')
* attr always has truthy value because of dict(message=...) so
```python
request.redirect("/shop/cart?%s" % werkzeug.url_encode(attr))